internal/controllers/machinedeployment/machinedeployment_rollout_ondelete.go

```diff
@@ -75,115 +81,109 @@ func (r *Reconciler) rolloutOnDelete(ctx context.Context, md *clusterv1.MachineD
 	return nil
 }
 
+// planOnDelete determine how to proceed with the rollout when using the OnDelete strategy if we are not yet at the desired state.
+func (p *rolloutPlanner) planOnDelete(ctx context.Context) error {
+	// Scale up, if we can.
+	if err := p.reconcileNewMachineSet(ctx); err != nil {
+		return err
+	}
+
+	// Scale down, if we can.
+	p.reconcileOldMachineSetsOnDelete(ctx)
+	return nil
+}
+
 // reconcileOldMachineSetsOnDelete handles reconciliation of Old MachineSets associated with the MachineDeployment in the OnDelete rollout strategy.
-func (r *Reconciler) reconcileOldMachineSetsOnDelete(ctx context.Context, oldMSs []*clusterv1.MachineSet, allMSs []*clusterv1.MachineSet, deployment *clusterv1.MachineDeployment) error {
+func (p *rolloutPlanner) reconcileOldMachineSetsOnDelete(ctx context.Context) {
 	log := ctrl.LoggerFrom(ctx)
-	if deployment.Spec.Replicas == nil {
-		return errors.Errorf("spec replicas for MachineDeployment %q/%q is nil, this is unexpected",
-			deployment.Namespace, deployment.Name)
+	oldMachinesCount := mdutil.GetReplicaCountForMachineSets(p.oldMSs)
+	if oldMachinesCount == 0 {
+		// Can't scale down further
+		return
 	}
-	log.V(4).Info("Checking to see if machines have been deleted or are in the process of deleting for old machine sets")
-	totalReplicas := mdutil.GetReplicaCountForMachineSets(allMSs)
-	scaleDownAmount := totalReplicas - *deployment.Spec.Replicas
-	for _, oldMS := range oldMSs {
-		log := log.WithValues("MachineSet", klog.KObj(oldMS))
-		if oldMS.Spec.Replicas == nil || *oldMS.Spec.Replicas <= 0 {
-			log.V(4).Info("fully scaled down")
+
+	// Determine if there are replicas to be scaled down due to a scale down in MD.
```

Comment:
This could be due to multiple reasons not just " due to a scale down in MD.", right?

EDIT: Hm or is this the only case where we can end up with more Machines than MD.spec.replicas?